cgifsave: avoid size issue

dloebl opened the pull request:
Avoid size issues when resizing GIF animations with a static background. The trick is to restore the original transparency setting after quantization and dithering.
**Background:** Keeping many pixels at the transparency index can reduce the size a lot and is the reason why the original animations in (#2576) are quite small in terms of bytes despite their huge width/height. The current resizing and subsequent color quantization method puts the focus on image quality but can make this transparency optimization less effective.

The suggested fix depends on dloebl/cgif#38 and cgif 0.2.0 to be released:

**Results:**
`$ vipsthumbnail --size 200 city.gif[n=-1] -o vips_city.gif`
`$ vipsthumbnail --size 1400 watch.gif[n=-1] -o vips_watch.gif`
`$ du -h vips_*`
**`2.2M vips_city.gif`
`812K vips_watch.gif`**

The size is still larger than before, as the transparency setting is changed by the resizing. However, the two animations are probably extreme cases and the size problem is mitigated quite a bit with this change.
I've visualized the transparency setting for the resized city GIF and the original (green means that a pixel is identical to the frame before):
**Resized:**
![out](https://user-images.githubusercontent.com/34104349/150117814-27d6c607-b24a-4f7e-8b76-159527b25619.gif)
**Original:**
![sorg](https://user-images.githubusercontent.com/34104349/150118018-6e6a1892-c096-4553-90a2-4bc96da2cf99.gif)

One can see some "kernels" (in the resized image). I guess these come from the interpolation method and illustrate where the remaining increase in size comes from. To further improve the size it could be interesting to have a "transparency-friendly" resize method (which keeps identical areas identical). For instance, nearest-neighbor resizing would be "transparency-friendly" but is of course not ideal in terms of quality. An alternative would be to determine transparent areas in a lossy way in cgifsave (I think @jcupitt once mentioned such a low-pass filter). So the problem is finding the right balance between quality and size.

Please let me know if there are changes required on this PR.

jcupitt commented:
> @jcupitt: Do you think such a modification would make sense? How would you weight the different color channels?

Yes, that's a good idea. CIELAB is the simplest perceptually uniform colour space (mostly uniform anyway!) so I would use that.

https://en.wikipedia.org/wiki/CIELAB_color_space#From_CIEXYZ_to_CIELAB

sRGB -> XYZ is this:

https://en.wikipedia.org/wiki/SRGB#From_CIE_XYZ_to_sRGB

You only need 8 bit data and you don't care about small errors, so you can cut a lot of corners. You'll find you end up with something like:

- Look up table for each of RGB to get linear data
- 3x3 matmul (so R' = a * R + b * G + c * B etc.)
- Look up table for cuberoot term in LAB
- L = G, a = 500 * (R - G), b = 200 * (G - B)

Not too painful.

jcupitt commented on libvips/foreign/cgifsave.c:
```diff
@@ -303,6 +335,36 @@ vips_foreign_save_cgif_write_frame( VipsForeignSaveCgif *cgif )
+				} else {
+					memcpy(cgif->frame_bytes_head + 4 * i, frame_bytes + 4 * i, 4);
+				}
+			}
```
The `*` inside the inner loop makes me uncomfortable, though the compiler ought to optimize this away. I think I'd still rework this as:

```C
                        const uint8_t trans_index = cgif->lp->count;
                        int n_pixels = frame_rect->width * frame_rect->height;
                        double maxerror = cgif->maxerror;

                        VipsPel *cur, *bef;
                        int i;

                        cur = frame_bytes;
                        bef = cgif->frame_bytes_head;
                        for( i = 0; i < n_pixels; i++ ) {
                                if( cgif_pixels_are_equal( cur, bef, maxerror ) )   
                                        cgif->index[i] = trans_index;
                                else {
                                        bef[0] = cur[0];
                                        bef[1] = cur[1];
                                        bef[2] = cur[2];
                                        bef[3] = cur[3];
                                }

                                cur += 4;
                                bef += 4;
                        }
```

Not all compilers will take out the very slow `memcpy()` for you.

You might as well make `maxerror` a double in the operation args as well.

jcupitt reviewed (commented):
The C programmer in me couldn't resist making some unnecessary optimisation, of course :( gif write is about 10% faster on my PC with the suggested changes. Maybe not a useful amount.

I noticed some small layout and naming things too.

dloebl commented:
Makes all perfect sense. Thanks!
I added your recommendations accordingly.
Additionally, I switched the lossy transparency off by default (maxerror = 0).
I think it makes sense to be a bit conservative before the pixel compare function has been improved (RGB to CIELAB).
